dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ConcurrentPriorityBlockingQueue.java

```diff
@@ -0,0 +1,666 @@
+     */
+    private final ReduceableSemaphore size = new ReduceableSemaphore(0, false);
+
+    /**
+     * The currentHead field is set to nonExpeditedHead initially until an expedited item is added to the queue
+     */
+    private final AtomicReference<Head<T>> currentHead = new AtomicReference<>();
+
+    @SuppressWarnings("unchecked")
+    private final FirstAction<T, Node<T>> GET_FIRST_NODE = GetFirstNode.INSTANCE;
+
+    @SuppressWarnings("unchecked")
+    private final FirstAction<T, T> GET_FIRST_ITEM = GetFirstItem.INSTANCE;
+
+    private final FirstAction<T, T> REMOVE_FIRST_ITEM;
+
+    ConcurrentPriorityBlockingQueue() {
+        nonExpeditedHead = new NonExpeditedHead<>();
+        nonExpeditedTail.set(nonExpeditedHead);
+        currentHead.set(nonExpeditedHead);
+
+        // There must be a node for the head and tail to point to
+        Node<T> exNode = new Node<>(null);
+        exNode.next.set(nonExpeditedHead);
+        expeditedHead = new ExpeditedHead<>(exNode);
+        expeditedTail.set(exNode);
+
+        REMOVE_FIRST_ITEM = new RemoveFirstItem<T>(nonExpeditedHead);
+    }
+
+    @Override
+    public boolean add(T e) {
+        return offer(e);
+    }
+
+    /**
+     * Determine if the provided Object is in the queue.
+     *
+     * If the Object is a QueueItem and it is marked expedited, look at the expedited part of the queue,
+     * otherwise look at the non-expedited part of the queue.
+     *
+     * @param item the item to determine if it is in the queue
+     */
+    @Override
+    public boolean contains(Object item) {
+        if (item == null) {
+            return false;
+        }
+
+        Head<T> head;
+        Node<T> end;
+
+        if (item instanceof QueueItem && ((QueueItem) item).isExpedited()) {
+            // If it is an expedited item, but no expedited items have been added, return false
+            if (currentHead.get() == nonExpeditedHead) {
+                return false;
+            }
+            head = expeditedHead;
+            end = nonExpeditedHead;
+        } else {
+            head = nonExpeditedHead;
+            end = null;
+        }
+
+        Node<T> current = getFirstWithAction(head, end, false, GET_FIRST_NODE);
+        for (; current != null; current = getNext(current, head, end)) {
+            T element = current.get();
+            if (element != null && item.equals(element)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private Node<T> getNext(Node<T> current, Head<T> head, Node<T> end) {
+        Node<T> next = current.next.get();
+        // If next points back to the current node, it means it was removed
+        // In that case start over again at the beginning
+        return (current == next) ? head.getFirst() : (next != end ? next : null);
+    }
+
+    @Override
+    public int drainTo(Collection<? super T> col) {
+        if (col == null) {
+            throw new NullPointerException();
```

Comment:
From the looks of it, the exceptions seem straightforward to debug, but figured I'd ask if there's any point in adding any message to these NPE or IAE? I can see not needing it, but just thinking of any serviceability gains

Reply:
These are the normal Exceptions from the other implementations.  These methods are actually called, but their implementations are here for completeness.  This class is used internally only, but I could update them to have a message if you think it has value.